transaction.py

```diff
@@ -295,3 +363,138 @@ def __bytes__(self):
             ret += self.bindingSig
 
         return ret
+
+
+class TransactionV5(object):
+    def __init__(self, rand, consensus_branch_id):
+        # Decide which transaction parts will be generated.
+        flip_coins = rand.u8()
+        have_transparent_in = (flip_coins >> 0) % 2
+        have_transparent_out = (flip_coins >> 1) % 2
+        have_sapling = (flip_coins >> 2) % 2
+        have_orchard = (flip_coins >> 3) % 2
+
+        # Common Transaction Fields
+        self.nVersionGroupId = NU5_VERSION_GROUP_ID
+        self.nConsensusBranchId = consensus_branch_id
+        self.nLockTime = rand.u32()
+        self.nExpiryHeight = rand.u32() % TX_EXPIRY_HEIGHT_THRESHOLD
+
+        # Transparent Transaction Fields
+        self.vin = []
+        self.vout = []
+        if have_transparent_in:
+            for _ in range((rand.u8() % 3) + 1):
+                self.vin.append(TxIn(rand))
+        if have_transparent_out:
+            for _ in range((rand.u8() % 3) + 1):
+                self.vout.append(TxOut(rand))
+
+        # Sapling Transaction Fields
+        self.vSpendsSapling = []
+        self.vOutputsSapling = []
+        if have_sapling:
+            self.anchorSapling = Fq(leos2ip(rand.b(32)))
+            for _ in range(rand.u8() % 3):
+                self.vSpendsSapling.append(SpendDescription(rand, self.anchorSapling))
+            for _ in range(rand.u8() % 3):
+                self.vOutputsSapling.append(OutputDescription(rand))
+            self.valueBalanceSapling = rand.u64() % (MAX_MONEY + 1)
+            self.bindingSigSapling = RedJubjubSignature(rand)
+        else:
+            # If valueBalanceSapling is not present in the serialized transaction, then
+            # v^balanceSapling is defined to be 0.
+            self.valueBalanceSapling = 0
+
+        # Orchard Transaction Fields
+        self.vActionsOrchard = []
+        if have_orchard:
+            for _ in range(rand.u8() % 5):
+                self.vActionsOrchard.append(OrchardActionDescription(rand))
+            self.flagsOrchard = rand.u8() & 3 # Only two flag bits are currently defined.
+            self.valueBalanceOrchard = rand.u64() % (MAX_MONEY + 1)
+            self.anchorOrchard = PallasBase(leos2ip(rand.b(32)))
+            self.proofsOrchard = rand.b(rand.u8() + 32) # Proof will always contain at least one element
+            self.bindingSigOrchard = RedPallasSignature(rand)
+        else:
+            # If valueBalanceOrchard is not present in the serialized transaction, then
+            # v^balanceOrchard is defined to be 0.
+            self.valueBalanceOrchard = 0
+
+    def header(self):
```

Comment:
This isn't exactly the transaction version, because it includes the "overwintered" flag bit. But it's also not the full transaction header anymore in the v5 format. Hmm...